### ISSUE TITLE
Ignore case new syntax

### ISSUE THREAD

**Issue by KvanTTT** [antlr/antlr4#1002]
What do you think about new lexer command for case ignoring?

``` ANTLR
Procedure: 'procedure' -> ignoreCase;
```

Or to use regex-like syntax (which gives more flexibility, for example ignore part of word):

``` ANTLR
Procedure: '\iprocedure\-i';
Procedure: '\i' 'procedure' '\-i';
```

Now I have to use the following not pretty syntax:

``` ANTLR
Procedure: P R O C E D U R E;
fragment A: [aA];
fragment B: [bB];
fragment C: [cC];
fragment D: [dD];
fragment E: [eE];
fragment F: [fF];
fragment G: [gG];
fragment H: [hH];
fragment I: [iI];
fragment J: [jJ];
fragment K: [kK];
fragment L: [lL];
fragment M: [mM];
fragment N: [nN];
fragment O: [oO];
fragment P: [pP];
fragment Q: [qQ];
fragment R: [rR];
fragment S: [sS];
fragment T: [tT];
fragment U: [uU];
fragment V: [vV];
fragment W: [wW];
fragment X: [xX];
fragment Y: [yY];
fragment Z: [zZ];
```

This feature especially important for case-insensivity languages, such as Pascal, PHP, etc.

**Comment by jimidle** [comment]
Don't create all to fragment rule just override LA in the input stream and always return toLower. 
Jim






On Mon, Sep 21, 2015 at 4:04 AM -0700, "Ivan Kochurkin" <notifications@github.com> wrote:












What do you think about new lexer command for case ignoring?

Procedure: 'procedure' -> ignoreCase;



Or to use regex-like syntax (which gives more flexibility, for example ignore part of word):

Procedure: '\iprocedure\-i';
Procedure: '\i' 'procedure' '\-i';



Now I have to use the following syntax:

Procedure: P R O C E D U R E;
fragment A: [aA];
fragment B: [bB];
fragment C: [cC];
fragment D: [dD];
fragment E: [eE];
fragment F: [fF];
fragment G: [gG];
fragment H: [hH];
fragment I: [iI];
fragment J: [jJ];
fragment K: [kK];
fragment L: [lL];
fragment M: [mM];
fragment N: [nN];
fragment O: [oO];
fragment P: [pP];
fragment Q: [qQ];
fragment R: [rR];
fragment S: [sS];
fragment T: [tT];
fragment U: [uU];
fragment V: [vV];
fragment W: [wW];
fragment X: [xX];
fragment Y: [yY];
fragment Z: [zZ];



This feature especially important for case-insensivity languages, such as Pascal, PHP, etc.

**Comment by KvanTTT** [comment]
>  Don't create all to fragment rule just override LA in the input stream and always return toLower.
@jimidle. Firstly, your solution runtime-dependent. Secondly, parsed language can be case-insensivity not for all parts (php).

There is such solution: https://github.com/developeron29/PLSQLParser/blob/master/PLSQL.g4#L534
But it is ugly, runtime-dependent (java) and, as I guess, it have not an optimal performance.

For example, the following tokens have an equal beginning, so it can be optimized.
```ANTLR
DO
    : D O
    ;

DOWNTO
    : D O W N T O
    ;
```

**Comment by jimidle** [comment]
It is faster than our fragment rules and as you specify only lower case letters in the lever rules it is better optimizes. I came up with original method and did the performance test. Regular expressions won't be implemented. So stick with your current method if you feel that overriding a single method is too much trouble.

**Comment by KvanTTT** [comment]
@jimidle , could you please show the grammar sample with your view of case insensitive tokens?

**Comment by jimidle** [comment]
I believe I have already put this in the public domain but of course your own needs may be different. 
Happy to share but here in Taiwan it is a public holiday so I will look again on Tuesday. 






On Fri, Sep 25, 2015 at 3:30 PM -0700, "Ivan Kochurkin" <notifications@github.com> wrote:












@jimidle , could you please show the grammar sample with your view of case insensitive tokens?

**Comment by KvanTTT** [comment]
Actually ```ignoreCase``` is not a lexer command, it is an option to perceive chars insensitive to case. So I would go for the following syntax (with respect to current ANTLR grammar):
```
Procedure: 'procedure' <ignoreCase=true>;
```
Or even easier:
```
Procedure: 'procedure' <ignoreCase>;
```

[Here](https://theantlrguy.atlassian.net/wiki/pages/viewpage.action?pageId=2687342) is described that "there is no ANTLR option that enables case insensitivity, as this is hard or impossible to do completely correctly, taking into account all possible internationalization issues", but in most cases not english chars are rarely used. Moreover it is possible to use warning "Lexer rule with not english chars and ignoreCase option may produce incorrect output". And for most cases performance with fragment rules is good. So, I think this option have the right to existence.

**Comment by yuriry** [comment]
I like <ignoreCase> option.  Not as convenient as in JavaCC' TOKEN[IGNORE_CASE] where multiple tokens can be marked as case-insensitive, but way better than various alternatives I've seen so far.

**Comment by KvanTTT** [comment]
Yes, the option is useful. But this issue a bit outdated. It's not possible to use option for a single token, but possible to use it for `mode` and entire grammar. See also PR with case insensitive proof of concept: https://github.com/antlr/antlr4/issues/1092.

**Comment by yuriry** [comment]
It seems that none of these options are available in the 4.4.6 release :(

**Comment by stefv** [comment]
I'm agree, it's a sad news because this feature can be usefull in the grammar to test immediatly with the tools without to create a program with a spacial input stream.

**Comment by KvanTTT** [comment]
Eventually fixed by #3399 and #3437
